datumaro/components/converter.py

```diff
@@ -1,21 +1,91 @@
-# Copyright (C) 2019-2021 Intel Corporation
+# Copyright (C) 2019-2022 Intel Corporation
 #
 # SPDX-License-Identifier: MIT
 
 from tempfile import mkdtemp
-from typing import Union
+from typing import NoReturn, Optional, Tuple, TypeVar, Union
 import logging as log
 import os
 import os.path as osp
 import shutil
 
+from attrs import define, field
+import attr
+
 from datumaro.components.cli_plugin import CliPlugin
-from datumaro.components.extractor import DatasetItem
+from datumaro.components.errors import (
+    AnnotationExportError, DatumaroError, ItemExportError,
+)
+from datumaro.components.extractor import DatasetItem, IExtractor
 from datumaro.components.media import Image
+from datumaro.components.progress_reporting import (
+    NullProgressReporter, ProgressReporter,
+)
 from datumaro.util.meta_file_util import save_meta_file
 from datumaro.util.os_util import rmtree
 from datumaro.util.scope import on_error_do, scoped
 
+T = TypeVar('T')
+
+class _ExportFail(DatumaroError):
+    pass
+
+class ExportErrorPolicy:
+    def report_item_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item.
+        If this function returns, the converter must skip the item.
+        """
+
+        if not isinstance(error, _ExportFail):
+            ie = ItemExportError(item_id)
+            ie.__cause__ = error
+            return self._handle_item_error(ie)
+        else:
+            raise error
+
+    def report_annotation_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item annotation.
+        If this function returns, the converter must skip the item.
+        """
+
+        if not isinstance(error, _ExportFail):
+            ie = AnnotationExportError(item_id)
+            ie.__cause__ = error
+            return self._handle_annotation_error(ie)
+        else:
+            raise error
+
+    def _handle_item_error(self, error: ItemExportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def _handle_annotation_error(self, error: AnnotationExportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def fail(self, error: Exception) -> NoReturn:
+        raise _ExportFail from error
+
+class FailingExportErrorPolicy(ExportErrorPolicy):
```

Comment:
Perhaps the failing behavior should just be the default? That way, a custom policy that wants to only ignore a certain kind of error will only need to override one method.

It'll also allow other kinds of errors to be added in the future without breaking compatibility.

Reply:
Yes, it makes sense. Done

Reply:
I don't think you pushed it.

Reply:
No, it's in 
[Make failing the default action](https://github.com/openvinotoolkit/datumaro/pull/650/commits/8736ea23284ef738767b31c6650b1283513c8d05)
[8736ea2](https://github.com/openvinotoolkit/datumaro/pull/650/commits/8736ea23284ef738767b31c6650b1283513c8d05)